optuna/_callbacks.py

```diff
@@ -97,21 +100,16 @@ def __init__(self, max_retry: Optional[int] = None) -> None:
         self._max_retry = max_retry
 
     def __call__(self, study: "optuna.study.Study", trial: FrozenTrial) -> None:
-        system_attrs = {"failed_trial": trial.number}
-
-        # Update the new object with the values in the trial.system_attrs.
-        # By doing this, if this failed try is already a rety, the 'failed_trial' value
-        # will be the first failed trial number.
-        system_attrs.update(trial.system_attrs)
-
-        retries = sum(
-            ("failed_trial", system_attrs["failed_trial"]) in t.system_attrs.items()
-            for t in study.trials
-        )
-
-        if self._max_retry is not None and retries + 1 > self._max_retry:
+        retry_count = trial.system_attrs.get("retry_count", 0) + 1
+        if self._max_retry is not None and retry_count > self._max_retry:
             return
 
+        system_attrs = {
+            **trial.system_attrs,
+            "failed_trial": trial.number,
+            "retry_count": retry_count,
+        }
+
```

Comment:
In the distributed case, if `retry_count` is referenced and updated in the following order, consistency may be lost.
Do you have any idea to resolve it? I'm not sure but introducing the lock will resolve it, or removing the new attribute `retru_count` will do.
```
process A: -----> R  --------------> U -------------->
process B: --------------> R --------------> U ------>
(R: reference the `retry_count` from RDB and U: update the `retry_count` in RDB)
```

Reply:
In my understanding the retry process should be run only in the master process in DDP setup.
https://github.com/optuna/optuna-examples/blob/29720558a2bac9e0bead3a237640f4013f4ee2bd/pytorch/pytorch_distributed_simple.py#L179-L187

More specifically, `study.optimize` calls [`fail_stale_trials()`](https://github.com/optuna/optuna/blob/v3.0.0-a0/optuna/study/_optimize.py#L192) -> [_heartbeat](https://github.com/optuna/optuna/blob/v3.0.0-a0/optuna/storages/_heartbeat.py#L27-L32) -> eventually reaches to `RetryFailedTrialCallback`, but as in the example `optimize()` is called only from the master, so basically we don't have to concern about updating information in an RDB of the same trial from multiple processes.
(If we call `optimize` in each worker, they'll have different params suggested)

When referring these values (`retry_count`, `failed_trial` etc) at the beginning of each trial as follows,
https://github.com/optuna/optuna-examples/blob/main/pytorch/pytorch_checkpoint.py#L87
[`TorchDistributedTrial.system_attrs`](https://github.com/optuna/optuna/blob/v3.0.0-a0/optuna/integration/pytorch_distributed.py#L224) gathers them from the master process through MPI broadcasting instead of directly accessing to DB, and MPI automatically synchronizes all the workers when a communication occurs, so I think there's basically no risk of inconsistency no matter the timing of reference.

Reply:
Thank you for the fruitful discussion. I understand the current code does not cause any inconsistency if we call the callback only in the master process. 

Should we add more documentation to the `RetryFailedTrialsCallback` which cannot be called in the multiple processes?

Reply:
Well I don't think so, because it'll never happen unless we try to call `study.optimize()` from multiple worker processes (in my understanding this is the case in MPI setup generally, but would there be any exceptions?).
So my opinion would be that it's rather worth mentioning it in the documentation of `Study`, maybe like "do not call `study.optimize()` in child workers because it'll make each worker behave inconsistently (due to being suggested different params worker-wise)", maybe with a reference to the [official example pytorch_distributed_simple.py](https://github.com/optuna/optuna-examples/blob/main/pytorch/pytorch_distributed_simple.py).
https://optuna.readthedocs.io/en/stable/reference/generated/optuna.study.Study.html

For users I think this should be OK, but for developers who works on integrating a new tool that uses MPI stuff to Optuna, I guess we need to somehow make them sure to follow how `TorchDistributedTrial` / `ChainerMNTrial` are implemented especially around their `system_attrs`.

Reply:
If you have multiple distributed processes processing a single trial, I don't think it will be a problem as you mentioned. On the other hand, if multiple distributed processes are processing different trials, there may be inconsistencies in referencing and updating the retry_count, right? Since it is often the case in Optuna that distributed processes handle separate trials that do not communicate with each other using MPI, etc., it is likely that this callback function will be called for a failed trial at the same time in some processes.

Reply:
I'm still wondering if the retry_count inconsistency can occur under normal Optuna workloads. I'm sorry if I misunderstood something.

Reply:
OK I see what you meant!
When running multiple trials in child worker processes separately without co-operating each other (maybe using MPI just as a bulk launcher, right? - this is equivalent to [Easy Parallelization](https://optuna.readthedocs.io/en/v3.0.0-a0/tutorial/10_key_features/004_distributed.html)), each process is in charge of an independent trial, meaning that each worker looks at the different DB entry, there will be no conflict.